### ISSUE TITLE
Add tests for kube-proxy --detect-local-mode configurations

### ISSUE THREAD

**Issue by wedaly** [kubernetes/kubernetes#108525]
### What would you like to be added?

Add tests to validate kube-proxy NAT rules for each --detect-local-mode configuration in kind clusters.

### Why is this needed?

We want to validate that iptables NAT rules work as expected for all supported --detect-local-mode configurations in a Kubernetes cluster.

* [KEP 2450](https://github.com/kubernetes/enhancements/tree/0e4d5df19d396511fe41ed0860b0ab9b96f46a2d/keps/sig-network/2450-Remove-knowledge-of-pod-cluster-CIDR-from-iptables-rules) proposed adding a --detect-local-mode flag to kube-proxy.
* [PR 87748](https://github.com/kubernetes/kubernetes/pull/87748) added mode ClusterCIDR
* [PR 88935](https://github.com/kubernetes/kubernetes/pull/88935) added mode NodeCIDR
* [PR 95400](https://github.com/kubernetes/kubernetes/pull/95400) is adding modes InterfacePrefix and BridgeInterfaceName

**Comment by wedaly** [comment]
I'm available to work on this

**Comment by wedaly** [comment]
I think there are two main scenarios to validate:
A) Packets between two pods through a service ClusterIP aren't NAT'd. Pretty sure this is already covered by [an existing test](https://github.com/kubernetes/kubernetes/blob/f93be6584ea57d1d01a0ca1d2c5e96782713d311/test/e2e/network/service.go#L928).
B) Packets from a pod to a node through a NodePort service are forwarded and NAT'd. I couldn't find a test that covers this, but I think it should be relatively straightforward to add.

We can then configure jobs to run these tests on a kind cluster, overriding the DetectLocalMode kube-proxy configuration (add KubeProxyConfiguration to kubeadmConfigPatches in the kind config generated by [e2e-k8s.sh](https://github.com/kubernetes-sigs/kind/blob/71af838275c3ba5b4c9663d0923a78738786a4b4/hack/ci/e2e-k8s.sh#L152))

**Comment by aojea** [comment]
I think that the tests are there

https://github.com/kubernetes/kubernetes/blob/f93be6584ea57d1d01a0ca1d2c5e96782713d311/test/e2e/network/networking.go#L143-L158

what is needed is to setup jobs with kube-proxy using different configurations, we have currently all this jobs with kind

https://testgrid.k8s.io/sig-network-kind

**Comment by wedaly** [comment]
For the test configurations, I was thinking of adding these:

1. ci-kubernetes-kind-detect-local-node-cidr
2. ci-kubernetes-kind-detect-local-bridge-interface-name (after #95400 merged)
3. ci-kubernetes-kind-detect-local-bridge-interface-name (after #95400 merged)

These would have the same configuration as [ci-kubernetes-kind-network](https://github.com/kubernetes/test-infra/blob/1efdc7f59247ca75b5e7beb4bde9739e51142ce7/config/jobs/kubernetes/sig-network/sig-network-kind.yaml#L156), except with [KUBE_PROXY_DETECT_LOCAL_MODE](https://github.com/kubernetes-sigs/kind/pull/2670) set to the appropriate detect-local-mode configuration.

Would also need to update KubeProxyConfiguration in kind's [e2e-k8s.sh](https://github.com/kubernetes-sigs/kind/blob/71af838275c3ba5b4c9663d0923a78738786a4b4/hack/ci/e2e-k8s.sh#L152) to set BridgeInterfaceName and InterfacePrefix to match the network interface names used by kind, but not until after #95400 lands.

How does that sound @aojea and @aramase ?

**Comment by aojea** [comment]
> the network interface names used by kind, but not until after #95400 lands.
> 
> How does that sound @aojea and @aramase ?

commented on #95400 , let's start in experiment mode

**Comment by wedaly** [comment]
I realized that the "BridgeInterfaceName" mode might be hard to validate in kind, since it doesn't seem to be creating a bridge interface. We can still validate the other modes though -- NodeCIDR and InterfacePrefix (with prefix "veth") both work.

**Comment by wedaly** [comment]
completed in [#25606](https://github.com/kubernetes/test-infra/pull/25606) and [#25743](https://github.com/kubernetes/test-infra/pull/25743)
